src/EventListener/DeserializeListener.php

```diff
@@ -64,32 +66,43 @@ public function onKernelRequest(GetResponseEvent $event)
         $request->attributes->set(
             'data',
             $this->serializer->deserialize(
-                $request->getContent(), $attributes['resource_class'], $this->getFormat($request), $context
+                $request->getContent(), $attributes['resource_class'], $format, $context
             )
         );
     }
 
     /**
-     * Extracts the format form the Content-Type header and fallback to the format of the response.
+     * Extracts the format from the Content-Type header and check that it is supported.
      *
      * @param Request $request
      *
+     * @throws NotAcceptableHttpException
+     *
      * @return string
      */
     private function getFormat(Request $request) : string
     {
-        $format = $request->getContentType();
-        if (null !== $format && isset($this->formats[$format])) {
-            return $format;
+        $contentType = $request->headers->get('CONTENT_TYPE');
+        if (null === $contentType) {
+            throw new NotAcceptableHttpException('The "Content-Type" header must exist.');
         }
 
-        $requestFormat = $request->getRequestFormat();
-        if (isset($this->formats[$requestFormat])) {
-            return $requestFormat;
-        }
+        $format = $request->getFormat($contentType);
+        if (!isset($this->formats[$format])) {
+            $supportedMimeTypes = [];
+            foreach ($this->formats as $mimeTypes) {
+                foreach ($mimeTypes as $mimeType) {
+                    $supportedMimeTypes[] = $mimeType;
+                }
+            }
 
-        reset($this->formats);
+            throw new NotAcceptableHttpException(sprintf(
+                'The content-type "%s" is not supported. Supported MIME types are "%s".',
+                $contentType,
+                implode('", "', $supportedMimeTypes)
```

Comment:
It needs a loop... Implode is quickest (we use it everywhere for this specific use case in Symfony).